src/mcpwm/comparator.rs

```diff
@@ -0,0 +1,116 @@
+use core::ptr;
+
+use esp_idf_sys::{
+    esp, mcpwm_cmpr_handle_t, mcpwm_comparator_config_t, mcpwm_comparator_config_t__bindgen_ty_1,
+    mcpwm_gen_t, mcpwm_generator_set_actions_on_compare_event, mcpwm_new_comparator,
+    mcpwm_oper_handle_t, mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_DOWN,
+    mcpwm_timer_direction_t_MCPWM_TIMER_DIRECTION_UP,
+};
+
+use super::generator::{CountingDirection, NoCmpMatchConfig, OnMatchCfg};
+
+trait ComparatorChannel {}
+
+pub struct CmpX;
+impl ComparatorChannel for CmpX {}
+
+pub struct CmpY;
+impl ComparatorChannel for CmpY {}
+
+pub trait OptionalCmp {
+    type OnMatchCfg: OnMatchCfg;
+    type Cfg: OptionalCmpCfg<Cmp = Self>;
+
+    fn get_comparator_mut(&mut self) -> Option<&mut Comparator>;
+}
+
+impl OptionalCmp for Comparator {
+    type OnMatchCfg = super::generator::CountingDirection;
+    type Cfg = ComparatorConfig;
+
+    fn get_comparator_mut(&mut self) -> Option<&mut Comparator> {
+        Some(self)
+    }
+}
+
+pub struct NoCmp;
+impl OptionalCmp for NoCmp {
+    type OnMatchCfg = super::generator::NoCmpMatchConfig;
+    type Cfg = NoCmpCfg;
+
+    fn get_comparator_mut(&mut self) -> Option<&mut Comparator> {
+        None
+    }
+}
+
+pub struct Comparator(pub(crate) mcpwm_cmpr_handle_t);
+
+impl Comparator {
+    pub(crate) unsafe fn configure(&mut self, gen: &mut mcpwm_gen_t, cfg: CountingDirection) {
+        // TODO: "must be terminated by MCPWM_GEN_TIMER_EVENT_ACTION_END()"
+        mcpwm_generator_set_actions_on_compare_event(
```

Comment:
[`mcpwm_generator_set_actions_on_compare_event`](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/mcpwm.html#_CPPv444mcpwm_generator_set_actions_on_compare_event18mcpwm_gen_handle_t32mcpwm_gen_compare_event_action_tz) is a variadic function that accepts a list of `mcpwm_gen_handle_t generator, mcpwm_gen_compare_event_action_t ev_act` pairs. Terminated by a special set of values.

Since we can only pass one single pair of values from rust this seems impossible, right?(unless the termination values is all we pass, which would be pointless)

Reply:
Looking at the implementation of [mcpwm_generator_set_actions_on_timer_event](https://github.com/espressif/esp-idf/blob/5f25bd843721dbddb76ab69099dc1b3f4ae5f759/components/driver/mcpwm/mcpwm_gen.c#L163) the functions seems to continue through the arguments until an `ev_act` with an `event = MCPWM_TIMER_EVENT_INVALID` is found. Looks like this is what [MCPWM_GEN_TIMER_EVENT_ACTION_END](https://github.com/espressif/esp-idf/blob/fde4afc67a2035a1f461dc511cd85dde1ef90368/components/driver/include/driver/mcpwm_gen.h#L87)

Same thing for [mcpwm_generator_set_actions_on_compare_event](https://github.com/espressif/esp-idf/blob/5f25bd843721dbddb76ab69099dc1b3f4ae5f759/components/driver/mcpwm/mcpwm_gen.c#L182) and `MCPWM_GEN_COMPARE_EVENT_ACTION_END()` but here a null pointer is used as `comparator`.

Still, we would need to be able to pass atleast two pairs of values(the real value pair and the end marker) to do something useful. I assume this is hard from rust?

Reply:
I have little to no knowledge on C variadic functions or ABIs, but would calling a function like
```rust
// Also, make sure bindgen does not generate bindings for the single `ec_act` version of this function
extern "C" fn mcpwm_generator_set_actions_on_compare_event(
    gen: mcpwm_gen_handle_t,
    ev_act0: mcpwm_gen_compare_event_action_t,
    ev_act1: mcpwm_gen_compare_event_action_t
) -> esp_err_t
```

just do the right thing and place the values in the places expected by the called function that actually is variadic?

Reply:
Yes I think it works like that. On the rust side you just call it with however many args you want and the compiler takes care of making sure the variadic ABI is correct.